api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java

```diff
@@ -455,70 +456,76 @@ protected boolean containsAlias(Iterator<CriteriaImpl.Subcriteria> subcriteriaIt
-						break;
+			handleAndListParam(locationReference, token -> {
+				if (token.getChain() != null) {
+					switch (token.getChain()) {
+						case Location.SP_ADDRESS_CITY:
+							propertyLike("l.cityVillage", token.getValue()).ifPresent(criteria::add);
+							break;
+						case Location.SP_ADDRESS_STATE:
+							propertyLike("l.stateProvince", token.getValue()).ifPresent(criteria::add);
+							break;
+						case Location.SP_ADDRESS_POSTALCODE:
+							propertyLike("l.postalCode", token.getValue()).ifPresent(criteria::add);
+							break;
+						case Location.SP_ADDRESS_COUNTRY:
+							propertyLike("l.country", token.getValue()).ifPresent(criteria::add);
+						case "":
+							criteria.add(eq("l.uuid", token.getValue()));
+							break;
+					}
 				}
-			}
+				return Optional.empty();
+			});
 		}
 	}
 	
-	protected void handleParticipantReference(Criteria criteria, ReferenceParam participantReference) {
+	protected void handleParticipantReference(Criteria criteria, ReferenceAndListParam participantReference) {
 		if (participantReference != null) {
 			criteria.createAlias("encounterProviders", "ep");
 			
-			if (participantReference.getChain() != null) {
-				switch (participantReference.getChain()) {
-					case Practitioner.SP_IDENTIFIER:
-						criteria.createAlias("ep.provider", "p").add(ilike("p.identifier", participantReference.getValue()));
-						break;
-					case Practitioner.SP_GIVEN:
-						criteria.createAlias("ep.provider", "pro").createAlias("pro.person", "ps")
-						        .createAlias("ps.names", "pn")
-						        .add(ilike("pn.givenName", participantReference.getValue(), MatchMode.START));
-						break;
-					case Practitioner.SP_FAMILY:
-						criteria.createAlias("ep.provider", "pro").createAlias("pro.person", "ps")
-						        .createAlias("ps.names", "pn")
-						        .add(ilike("pn.familyName", participantReference.getValue(), MatchMode.START));
-						break;
-					case Practitioner.SP_NAME:
-						criteria.createAlias("ep.provider", "pro").createAlias("pro.person", "ps").createAlias("ps.names",
-						    "pn");
-						
-						List<Optional<Criterion>> criterionList = new ArrayList<>();
-						
-						for (String token : StringUtils.split(participantReference.getValue(), " \t,")) {
-							criterionList.add(propertyLike("pn.givenName", token));
-							criterionList.add(propertyLike("pn.middleName", token));
-							criterionList.add(propertyLike("pn.familyName", token));
-						}
-						
-						criteria.add(or(toCriteriaArray(criterionList)));
-						break;
-					case "":
-						criteria.add(eq("ep.uuid", participantReference.getValue()));
-						break;
+			handleAndListParam(participantReference, participantToken -> {
+				if (participantToken.getChain() != null) {
+					switch (participantToken.getChain()) {
+						case Practitioner.SP_IDENTIFIER:
+							criteria.createAlias("ep.provider", "p").add(ilike("p.identifier", participantToken.getValue()));
```

Comment:
When we're doing this inside a `handleAndListParam` we do need to check that the alias doesn't already exist.